File: app-android/app/src/main/java/com/aengix/hopper/tunnel/IPTunnelEngine.kt
package com.aengix.hopper.tunnel

import com.aengix.hopper.ssh.SSHByteStream
import com.aengix.hopper.util.HopErrorDetails
import com.aengix.hopper.util.TunnelLog
import java.io.FileInputStream
import java.io.FileOutputStream
import java.util.concurrent.atomic.AtomicBoolean

class IPTunnelEngine(
    private val stream: SSHByteStream,
    private val tunInput: FileInputStream,
    private val tunOutput: FileOutputStream,
) {
    private val running = AtomicBoolean(false)
    private val threads = mutableListOf<Thread>()
    var onFailure: ((String) -> Unit)? = null

    @Volatile
    private var pendingInbound = ByteArray(0)
    private val bufferLock = Any()

    fun start() {
        if (!running.compareAndSet(false, true)) return
        TunnelLog.info("IPTunnelEngine starting")
        threads += startThread("tun-to-ssh") { tunToSSH() }
        threads += startThread("ssh-to-tun") { sshToTun() }
        threads += startThread("keepalive") { keepaliveLoop() }
    }

    fun stop() {
        if (!running.compareAndSet(true, false)) return
        threads.forEach { it.interrupt() }
        threads.clear()
        stream.close()
        TunnelLog.info("IPTunnelEngine stopped")
    }

    private fun tunToSSH() {
        val buffer = ByteArray(32_767)
        while (running.get() && !Thread.currentThread().isInterrupted) {
            try {
                val read = tunInput.read(buffer)
                if (read <= 0) {
                    Thread.sleep(5)
                    continue
                }
                val frame = IPTunnelFrame(IPTunnelFrameType.DATA, buffer.copyOf(read))
                stream.write(frame.encoded())
            } catch (error: Exception) {
                if (!running.get()) return
                fail("TUN write to SSH failed: ${HopErrorDetails.describe(error)}")
                return
            }
        }
    }

    private fun sshToTun() {
        val readBuffer = ByteArray(32_768)
        while (running.get() && !Thread.currentThread().isInterrupted) {
            try {
                val read = stream.read(readBuffer, 0, readBuffer.size)
                if (read < 0) {
                    if (!running.get()) return
                    fail("SSH read failed: stream closed")
                    return
                }
                if (read == 0) continue

                synchronized(bufferLock) {
                    pendingInbound = pendingInbound + readBuffer.copyOf(read)
                    processInbound()
                }
            } catch (error: Exception) {
                if (!running.get()) return
                fail("SSH read failed: ${HopErrorDetails.describe(error)}")
                return
            }
        }
    }

    private fun processInbound() {
        var cursor = pendingInbound
        while (true) {
            if (cursor.size < IPTunnelFrame.HEADER_LENGTH) {
                pendingInbound = cursor
                return
            }

            if (cursor[0] != 1.toByte()) {
                fail("Invalid iptunnel frame: badVersion (head=${hexPreview(cursor)})")
                return
            }

            if (IPTunnelFrameType.fromRaw(cursor[1]) == null) {
                fail("Invalid iptunnel frame: badType (head=${hexPreview(cursor)})")
                return
            }

            val payloadLength = ((cursor[2].toInt() and 0xFF) shl 8) or (cursor[3].toInt() and 0xFF)
            if (payloadLength > IPTunnelFrame.MAX_PACKET_LENGTH) {
                fail("Invalid iptunnel frame: packetTooLarge ($payloadLength)")
                return
            }

            val frameLength = IPTunnelFrame.HEADER_LENGTH + payloadLength
            if (cursor.size < frameLength) {
                pendingInbound = cursor
                return
            }

            val frameData = cursor.copyOfRange(0, frameLength)
            cursor = cursor.copyOfRange(frameLength, cursor.size)

            try {
                val frame = IPTunnelFrame.decode(frameData)
                when (frame.type) {
                    IPTunnelFrameType.DATA -> {
                        tunOutput.write(frame.payload)
                        tunOutput.flush()
                    }
                    IPTunnelFrameType.KEEPALIVE -> Unit
                }
            } catch (error: Exception) {
                fail("Invalid iptunnel frame: ${error.message} (head=${hexPreview(frameData)})")
                return
            }
        }
    }

    private fun keepaliveLoop() {
        while (running.get() && !Thread.currentThread().isInterrupted) {
            try {
                Thread.sleep(25_000)
                if (!running.get()) return
                stream.write(IPTunnelFrame(IPTunnelFrameType.KEEPALIVE).encoded())
            } catch (error: Exception) {
                if (!running.get()) return
                fail("SSH keepalive failed: ${HopErrorDetails.describe(error)}")
                return
            }
        }
    }

    private fun fail(message: String) {
        TunnelLog.error(message)
        onFailure?.invoke(message)
        stop()
    }

    private fun hexPreview(data: ByteArray, limit: Int = 16): String =
        data.take(limit).joinToString(" ") { "%02x".format(it) }

    private fun startThread(name: String, block: () -> Unit): Thread =
        Thread(block, name).also { it.start() }
}

private operator fun ByteArray.plus(other: ByteArray): ByteArray =
    ByteArray(size + other.size).also { result ->
        copyInto(result)
        other.copyInto(result, size)
    }

File: app/TunnelCore/IPTunnelEngine.swift
import Darwin
import Foundation
import NetworkExtension

final class IPTunnelEngine {
    private let stream: SSHByteStream
    private let packetIO: PacketFlowIO
    private var tasks: [Task<Void, Never>] = []
    private var inboundBuffer = Data()
    private let bufferLock = NSLock()

    var onFailure: (@Sendable (String) -> Void)?

    init(stream: SSHByteStream, packetFlow: NEPacketTunnelFlow) {
        self.stream = stream
        self.packetIO = PacketFlowIO(packetFlow: packetFlow)
    }

    func start() {
        TunnelLog.info("IPTunnelEngine starting")
        tasks.append(Task { await self.tunToSSH() })
        tasks.append(Task { await self.sshToTun() })
        tasks.append(Task { await self.keepaliveLoop() })
    }

    func stop() {
        tasks.forEach { $0.cancel() }
        tasks.removeAll()
        stream.close()
        TunnelLog.info("IPTunnelEngine stopped")
    }

    private func tunToSSH() async {
        while !Task.isCancelled {
            let (packets, _) = await packetIO.readPackets()
            if Task.isCancelled { return }
            if packets.isEmpty {
                try? await Task.sleep(nanoseconds: 5_000_000)
                continue
            }
            for packet in packets {
                guard !packet.isEmpty else { continue }
                let frame = IPTunnelFrame(type: .data, payload: packet)
                do {
                    try await stream.write(frame.encoded())
                } catch {
                    fail("TUN write to SSH failed: \(HopErrorDetails.describe(error))")
                    return
                }
            }
        }
    }

    private func sshToTun() async {
        while !Task.isCancelled {
            let chunk: Data
            do {
                chunk = try await stream.read()
            } catch {
                if Task.isCancelled { return }
                fail("SSH read failed: \(HopErrorDetails.describe(error))")
                return
            }
            if chunk.isEmpty {
                continue
            }

            bufferLock.lock()
            inboundBuffer.append(chunk)
            let buffer = inboundBuffer
            inboundBuffer = Data()
            bufferLock.unlock()

            var cursor = buffer
            while true {
                guard cursor.count >= IPTunnelFrame.headerLength else {
                    bufferLock.lock()
                    inboundBuffer = cursor
                    bufferLock.unlock()
                    break
                }

                guard cursor[0] == IPTunnelFrame.wireVersion else {
                    fail("Invalid iptunnel frame: badVersion (stream misaligned, head=\(Self.hexPreview(cursor)))")
                    return
                }

                guard IPTunnelFrameType(rawValue: cursor[1]) != nil else {
                    fail("Invalid iptunnel frame: badType (head=\(Self.hexPreview(cursor)))")
                    return
                }

                let payloadLength = Int(cursor.uint16BE(at: 2))
                guard payloadLength <= IPTunnelFrame.maxPacketLength else {
                    fail("Invalid iptunnel frame: packetTooLarge (\(payloadLength))")
                    return
                }

                let frameLength = IPTunnelFrame.headerLength + payloadLength
                guard cursor.count >= frameLength else {
                    bufferLock.lock()
                    inboundBuffer = cursor
                    bufferLock.unlock()
                    break
                }

                let frameData = cursor.prefix(frameLength)
                cursor = Data(cursor.dropFirst(frameLength))

                do {
                    let frame = try IPTunnelFrame.decode(from: frameData)
                    switch frame.type {
                    case .data:
                        packetIO.writePackets([frame.payload], withProtocols: [AF_INET as NSNumber])
                    case .keepalive, .assignReq, .assignResp:
                        break
                    }
                } catch {
                    fail("Invalid iptunnel frame: \(error) (head=\(Self.hexPreview(frameData)))")
                    return
                }
            }
        }
    }

    private func keepaliveLoop() async {
        while !Task.isCancelled {
            try? await Task.sleep(nanoseconds: 25_000_000_000)
            if Task.isCancelled { return }
            let frame = IPTunnelFrame(type: .keepalive)
            do {
                try await stream.write(frame.encoded())
            } catch {
                if Task.isCancelled { return }
                fail("SSH keepalive failed: \(HopErrorDetails.describe(error))")
                return
            }
        }
    }

    private func fail(_ message: String) {
        TunnelLog.error(message)
        onFailure?(message)
        stop()
    }

    private static func hexPreview(_ data: Data, limit: Int = 16) -> String {
        data.prefix(limit).map { String(format: "%02x", $0) }.joined(separator: " ")
    }
}
